code/modules/wiremod/shell/module.dm

```diff
@@ -0,0 +1,154 @@
+		var/action_name = action_component.button_name.value
+		if(!repeat_name_counts[action_name])
+			repeat_name_counts[action_name] = 0
+		repeat_name_counts[action_name]++
+		if(repeat_name_counts[action_name] > 1)
+			action_name += " ([repeat_name_counts[action_name]])"
+		display_names[action_name] = REF(action_component)
+		var/option_icon_state = "bci_[replacetextEx(lowertext(action_component.icon_options.value), " ", "_")]"
+		radial_options += list("[action_name]" = image('icons/mob/actions/actions_items.dmi', option_icon_state))
+	radial_options += list("Access Circuit" = image(shell.attached_circuit))
+	var/selected_option = show_radial_menu(mod.wearer, src, radial_options, custom_check = FALSE, require_near = TRUE)
+	if(!selected_option)
+		return
+	if(!mod || !mod.wearer || !mod.active || mod.activating)
+		return
+	if(selected_option == "Access Circuit")
+		shell.attached_circuit?.interact(mod.wearer)
+	else
+		var/component_reference = display_names[selected_option]
+		var/obj/item/circuit_component/equipment_action/mod/selected_component = locate(component_reference) in shell.attached_circuit.contents
+		if(!istype(selected_component))
+			return
+		selected_component.signal.set_output(COMPONENT_SIGNAL)
+
+
+/obj/item/circuit_component/mod_adapter_core
+	display_name = "MOD circuit adapter core"
+	desc = "Provides a reference to the MODsuit's occupant and allows the circuit to toggle the MODsuit."
+
+	/// The MODsuit module this circuit is associated with
+	var/obj/item/mod/module/attached_module
+
+	/// The name of the module to select
+	var/datum/port/input/module_to_select
+
+	/// The signal to toggle the suit
+	var/datum/port/input/toggle_suit
+
+	/// The signal to select a module
+	var/datum/port/input/select_module
+
+	/// A reference to the wearer of the MODsuit
+	var/datum/port/output/wearer
+
+	/// The name of the last selected module
+	var/datum/port/output/selected_module
+
+	/// The signal that is triggered when a module is selected
+	var/datum/port/output/on_module_selected
+
+/obj/item/circuit_component/mod_adapter_core/populate_ports()
+	// Input Signals
+	module_to_select = add_input_port("Module to Select", PORT_TYPE_STRING)
+	toggle_suit = add_input_port("Toggle Suit", PORT_TYPE_SIGNAL)
+	select_module = add_input_port("Select Module", PORT_TYPE_SIGNAL)
+	// States
+	wearer = add_output_port("Wearer", PORT_TYPE_ATOM)
+	selected_module = add_output_port("Selected Module", PORT_TYPE_STRING)
+	// Output Signals
+	on_module_selected = add_output_port("On Module Selected", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/mod_adapter_core/register_shell(atom/movable/shell)
+	if(istype(shell, /obj/item/mod/module))
+		attached_module = shell
+	RegisterSignal(attached_module, COMSIG_MOVABLE_MOVED, .proc/on_move)
+
+/obj/item/circuit_component/mod_adapter_core/unregister_shell(atom/movable/shell)
+	UnregisterSignal(attached_module, COMSIG_MOVABLE_MOVED)
+	attached_module = null
```

Comment:
ideally all components should call parent here just in case any sort of signal needs to be registered for when a shell gets unregistered/register